pkg/operator/clustermemberremovalcontroller/clustermemberremovalcontroller.go

```diff
@@ -0,0 +1,251 @@
+//
+// Note:
+//  since this controller needs to reconcile only master nodes and machine objects
+//  make sure nodeInformer and machineInformer contain only filtered data
+//  otherwise it might be expensive to react to every node update in larger installations
+func NewClusterMemberRemovalController(
+	etcdClient etcdcli.EtcdClient,
+	isMachineAPIFunctional func() (bool, error),
+	masterMachineSelector labels.Selector, masterNodeSelector labels.Selector,
+	kubeInformersForNamespaces operatorv1helpers.KubeInformersForNamespaces,
+	masterNodeInformer cache.SharedInformer,
+	masterMachineInformer cache.SharedInformer,
+	masterNodeLister corev1listers.NodeLister,
+	masterMachineLister machinelistersv1beta1.MachineLister,
+	hasNetworkInformerSyncedFn func() bool,
+	networkLister configv1listers.NetworkLister,
+	eventRecorder events.Recorder,
+) factory.Controller {
+	c := &ClusterMemberRemovalController{
+		etcdClient:                        etcdClient,
+		isMachineAPIFunctionalFn:          isMachineAPIFunctional,
+		masterMachineSelector:             masterMachineSelector,
+		masterNodeSelector:                masterNodeSelector,
+		masterNodeLister:                  masterNodeLister,
+		masterMachineLister:               masterMachineLister,
+		networkLister:                     networkLister,
+		configMapListerForTargetNamespace: kubeInformersForNamespaces.InformersFor(operatorclient.TargetNamespace).Core().V1().ConfigMaps().Lister().ConfigMaps(operatorclient.TargetNamespace),
+		hasNetworkInformerSyncedFn:        hasNetworkInformerSyncedFn,
+	}
+	return factory.New().
+		WithSync(c.sync).
+		ResyncEvery(33*time.Minute). // make it slow since nodes are updated every few minutes
+		WithInformers(
+			masterNodeInformer,
+			masterMachineInformer,
+			kubeInformersForNamespaces.InformersFor(operatorclient.TargetNamespace).Core().V1().ConfigMaps().Informer(),
+		).ToController("ClusterMemberRemovalController", eventRecorder.WithComponentSuffix("cluster-member-removal-controller"))
+}
+
+func (c *ClusterMemberRemovalController) sync(ctx context.Context, _ factory.SyncContext) error {
+	if !c.hasNetworkInformerSyncedFn() {
+		return nil
+	}
+	// stop if the machine API is not functional
+	if isFunctional, err := c.isMachineAPIFunctionalFn(); err != nil {
+		return err
+	} else if !isFunctional {
+		return nil
+	}
+	etcdEndpointsConfigMap, err := c.configMapListerForTargetNamespace.Get("etcd-endpoints")
+	if err != nil {
+		return err // should not happen
+	}
+	for len(etcdEndpointsConfigMap.Data) == 0 {
+		return nil
+	}
+	for _, potentialMemberToRemoveIP := range etcdEndpointsConfigMap.Data {
+		memberLocator := potentialMemberToRemoveIP
+		machine, err := c.getMachineForMember(potentialMemberToRemoveIP)
+		if err != nil {
+			klog.Errorf("unable to get a machine for member: %v, err: %v", memberLocator, err)
+			continue
```

Comment:
Shouldn't we be retrying here? I understand we want to loop over to the remaining member IPs but these errors might just be transient failures to GET/LIST machines or nodes.
Similar to how we're retrying in L128.
```suggestion
			return fmt.Errorf("unable to get a machine for member: %v, err: %v", memberLocator, err)
```
Same for L116 and L146 (which is a failure to remove the etcd member).

Reply:
I don't think we should retry, primarily because all calls read data from the internal cache - we won't issue a network call until L126

Reply:
Controller sync loops should return errors and get rekicked.  Even if a few particular cases may be ok without them, you're not expected a failure and it should be reported and retried.  Please return the error.